compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -259,6 +260,44 @@ impl<'a, 'tcx> CfgSimplifier<'a, 'tcx> {
     }
 }
 
+pub fn remove_duplicate_unreachable_blocks<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    struct OptApplier<'tcx> {
+        tcx: TyCtxt<'tcx>,
+        first: BasicBlock,
+        duplicates: FxHashSet<BasicBlock>,
+    }
+
+    impl<'tcx> MutVisitor<'tcx> for OptApplier<'tcx> {
+        fn tcx(&self) -> TyCtxt<'tcx> {
+            self.tcx
+        }
+
+        fn visit_terminator(&mut self, terminator: &mut Terminator<'tcx>, location: Location) {
+            for target in terminator.successors_mut() {
+                if self.duplicates.contains(target) {
+                    *target = self.first;
+                }
+            }
+
+            self.super_terminator(terminator, location);
+        }
+    }
+
+    let unreachable_blocks = body
+        .basic_blocks
+        .iter_enumerated()
+        .filter_map(|(block, bb)| {
+            if bb.terminator.is_some() && bb.is_empty_unreachable() { Some(block) } else { None }
+        })
+        .collect::<Vec<_>>();
+
+    if unreachable_blocks.len() > 1 {
+        let first = unreachable_blocks[0];
+        let duplicates = unreachable_blocks.into_iter().skip(1).collect();
```

Comment:
Can we avoid re-allocating by making `unreachable_blocks` a `FxIndexSet` from the start?